Topology management doc updates (#17451)
kubernetes/website: content/en/docs/tasks/administer-cluster/topology-manager.md

The Topology Manager currently:

 - Works on Nodes with the `static` CPU Manager Policy enabled. See [control CPU Management Policies](/docs/tasks/administer-cluster/cpu-management-policies/)

>  - Works on Pods making CPU or Device requests via extended resources

Reviewer: ```suggestion
 - Works on Pods making CPU requests or Device requests via extended resources
```

The Topology Manager currently:

 - Works on Nodes with the `static` CPU Manager Policy enabled. See [control CPU Management Policies](/docs/tasks/administer-cluster/cpu-management-policies/)

>  - Works on Pods making CPU or Device requsts via extended resources

Reviewer 2: ```suggestion
 - Works on Pods making CPU or Device requests via extended resources
```

preferred NUMA Node affinity for that container. If the affinity is not preferred, 
Topology Manager will reject this pod from the node. This will result in a pod in a `Terminated` state with a pod admission failure.

Once the pod is in a `Terminated` state, the Kubernetes scheduler will **not** attempt to reschedule the pod. It is recommended to use a ReplicaSet or Deployment to trigger a redeploy of the pod.

> An external control loop could be also implemented to trigger a redeployment of pods that have the `Topology Affiniy` error.

Reviewer: ```suggestion
An external control loop could be also implemented to trigger a redeployment of pods that have the `Topology Affinity` error.
```

preferred NUMA Node affinity for that container. If the affinity is not preferred, 
Topology Manager will reject this pod from the node. This will result in a pod in a `Terminated` state with a pod admission failure.

> Once the pod is a `Terminated` state, the Kubernetes scheduler will **not** attempt to reschedule the pod. It it recommended to use a ReplicaSet or Deployment to trigger a redeploy of the pod.

Reviewer 3: s/is a `Terminated` state/is in a `Terminated` state/

> Once the pod is a `Terminated` state, the Kubernetes scheduler will **not** attempt to reschedule the pod. It it recommended to use a ReplicaSet or Deployment to trigger a redeploy of the pod.

Reviewer 3: s/It it/It is/

resource allocation decision.
If, however, this is not possible then the Topology Manager will reject the pod from the node. This will result in a pod in a `Terminated` state with a pod admission failure.

Once the pod is in a `Terminated` state, the Kubernetes scheduler will **not** attempt to reschedule the pod. It is recommended a Deployment with Replicas to trigger a redeploy of the pod.

> An external control loop could be also implemented to trigger a redeployment of pods that have the `Topology Affiniy` error.

Reviewer: ```suggestion
An external control loop could be also implemented to trigger a redeployment of pods that have the `Topology Affinity` error.
```

resource allocation decision.
If, however, this is not possible then the Topology Manager will reject the pod from the node. This will result in a pod in a `Terminated` state with a pod admission failure.

> Once the pod is a `Terminated` state, the Kubernetes scheduler will **not** attempt to reschedule the pod. It it recommended a Deployment with Replicas to trigger a redeploy of the pod.

Reviewer 3: s/is a Terminated state/is in a Terminated state/

s/It it/It is/

The Topology Manager would consider both of the above pods. The Topology Manager would consult the Hint Providers, which are CPU and Device Manager to get topology hints for the pods. 
In the case of the `Guaranteed` pod the `static` CPU Manager policy would return hints relating to the CPU request and the Device Manager would send back hints for the requested device.

In the case of the `BestEffort` pod the CPU Manager would send back the default hint as there is no CPU request and the Device Manager would send back the hints for each of the requested devices.

Using this information the Topology Manager calculates the optimal hint for the pod and stores this information, which will be used by the Hint Providers when they are making their resource assignments. 

### Known Limitations
1. As of K8s 1.16 the Topology Manager is currently only guaranteed to work if a *single* container in the pod spec requires aligned resources. This is due to the hint generation being based on current resource allocations, and all containers in a pod generate hints before any resource allocation has been made. This results in unreliable hints for all but the first container in a pod.

> 

Reviewer: The scheduler is not topology-aware, so it's possible to pass the scheduler and fail the Admit() check in kubelet.  If a higher-level controller is used (replicaset, for example) it can repeatedly re-create the pod and have it schedule and fail the same way.

If multiple pods/containers are considered by kubelet in close succession, they can result in the topology manager policy being effectively ignored.  See https://github.com/kubernetes/kubernetes/issues/84749

```yaml
spec:
  containers:
  - name: nginx
    image: nginx
    resources:
      limits:
        example.com/deviceA: "1"
        example.com/deviceB: "1"
      requests:
        example.com/deviceA: "1"
        example.com/deviceB: "1"
```
This pod runs in the `BestEffort` QoS class because there are no CPU and memory requests.

The Topology Manager would consider both of the above pods. The Topology Manager would consult the Hint Providers, which are CPU and Device Manager to get topology hints for the pods. 
In the case of the `Guaranteed` pod the `static` CPU Manager policy would return hints relating to the CPU request and the Device Manager would send back hints for the requested device.

In the case of the `BestEffort` pod the CPU Manager would send back the default hint as there is no CPU request and the Device Manager would send back the hints for each of the requested devices.

Using this information the Topology Manager calculates the optimal hint for the pod and stores this information, which will be used by the Hint Providers when they are making their resource assignments. 

### Known Limitations

> 1. As of K8s 1.16 the Topology Manager is currently only guaranteed to work if a *single* container in the pod spec requires aligned resources. This is due to the hint generation being based on current resource allocations, and all containers in a pod generate hints before any resource allocation has been made. This results in unreiliable hints for all but the first contaienr in a pod.

Reviewer 2: ```suggestion
1. As of K8s 1.16 the Topology Manager is currently only guaranteed to work if a *single* container in the pod spec requires aligned resources. This is due to the hint generation being based on current resource allocations, and all containers in a pod generate hints before any resource allocation has been made. This results in unreliable hints for all but the first container in a pod.
```

This pod runs in the `Guaranteed` QoS class because `requests` are equal to `limits`.

Topology Manager would consider this Pod. The Topology Manager consults the CPU Manager `static` policy, which returns the topology of available CPUs. 
Topology Manager also consults the Device Manager to discover the topology of available devices for example.com/device.

Topology Manager will use this information to store the best Topology for this container. In the case of this Pod, CPU and Device Manager will use this stored information at the resource allocation stage.

```yaml
spec:
  containers:
  - name: nginx
    image: nginx
    resources:
      limits:
        example.com/deviceA: "1"
        example.com/deviceB: "1"
      requests:
        example.com/deviceA: "1"
        example.com/deviceB: "1"
```

> This pod runs in the `BestEffort` QoS class because there are not CPU and memory requests.

Reviewer 3: s/not/no/

> Topology Manager would consider this Pod. The Topology Manager consults the Device Manager to discover the topology of the available devices for example.com/deviceA and example.com/deviceB.

Reviewer 3: The flow to this sentence is weird from the sentence above.

As above Topology Manager will use this information to store the best Topology for this container. Device Manager will then use this when assigning devices to the Pod.

> {{% /capture %}}

Reviewer 3: We should probably add a section about known issues.

Author: Added the ones I could remember off hand, let me know what others I have missed
